### PR TITLE
Fix build_laplace_objective behaviour

### LINKED ISSUE #109: Laplace Approximation intermediates
As noted in the code, this line makes quite strong assumptions about `x` https://github.com/JuliaGaussianProcesses/ApproximateGPs.jl/blob/bae0a1ebe1865a9eba572ea86a7adbc75cd79387/src/laplace.jl#L60 .

When I went to use the laplace approximation code I initially stumbled around with it for a while, before realising I could just call `build_laplace_objective!`, and build `f` myself.

I'm not entirely sure what a proper fix that gives the correct initialisation all of the time would look like, but I think we can tell when it's definitely going to produce an error downstream, because I'm pretty sure we'll get an error if the element type of `xs` isn't a subtype of `Real`? Should we perhaps check this case, and provide a useful error message if it's hit?

@st-- ?

### PR DESCRIPTION
**Pull request by st--**
Resolves #109

### CODE REVIEW

**Review by willtebbutt — APPROVED** [review]
LGTM subject to patch bump and CI passing (I'm convinced by the test)

File: test/LaplaceApproximationModule.jl [hunk 1]
```diff
@@ -264,4 +264,13 @@
...
         @test res.q isa MvNormal
     end
+
+    @testset "GitHub issue #109" begin
+        build_latent_gp() = LatentGP(GP(SEKernel()), BernoulliLikelihood(), 1e-8)
+
+        x = ColVecs(randn(2, 5))
+        _, y = rand(build_latent_gp()(x))
+
+        build_laplace_objective(build_latent_gp, x, y)()
```
Comment (rossviljoen) [inline test/LaplaceApproximationModule.jl:274]: Maybe test that the produced objective function works as well?

Comment (st--) [reply test/LaplaceApproximationModule.jl:274]: it is actually called:) but I'll make that clearer

**Comment by willtebbutt** [comment]
Hmmm a good place to start might be to figure out where the `Core.Box` is coming from, and why its length is being asked for. It doesn't look obviously like an AD-specific thing to me, although perhaps it is if things that aren't using AD are working properly?

**Comment by st--** [comment]
>Hmmm a good place to start might be to figure out where the Core.Box is coming from, and why its length is being asked for. It doesn't look obviously like an AD-specific thing to me, although perhaps it is if things that aren't using AD are working properly?

That I can tell you: it's coming from the `objective` closure within `build_laplace_objective!`, because the externally-scoped `f` doesn't have a defined type, as it's `nothing` when first called and then `Vector` in subsequent calls (once it was first assigned in the line I added).

**Comment by st--** [comment]
> Hmm okay. So is a `Core.Box` making its way into the call to `map` before it gets assigned then? So it this actually an AD issue, or is it just to do with ensuring that `f` is initialised by the time it's needed?

I'm getting less sure here, but it's definitely not an issue in the forward pass (and all the `Box`es seem to get hidden away), but AD fails...

File: src/LaplaceApproximationModule.jl [hunk 2]
```diff
@@ -75,13 +75,29 @@ closure passes its arguments to `build_latent_gp`, which must return the
...
 
 function build_laplace_objective!(
-    f,
+    f_init::Vector,
+    build_latent_gp,
+    xs,
+    ys;
+    kwargs...
+)
+    return build_laplace_objective!(LaplaceObjectiveCache(f_init), build_latent_gp, xs, ys; kwargs...)
```
Comment (st--) [inline src/LaplaceApproximationModule.jl:92]:
```suggestion
    return build_laplace_objective!(
        LaplaceObjectiveCache(f_init), build_latent_gp, xs, ys; kwargs...
    )
```

File: src/LaplaceApproximationModule.jl [hunk 3]
```diff
@@ -75,13 +75,31 @@ closure passes its arguments to `build_latent_gp`, which must return the
...
 end
 
 function build_laplace_objective!(
-    f,
+    f_init::Vector,
+    build_latent_gp,
+    xs,
+    ys;
+    kwargs...
+)
```
[the remark this answers is not in the document]
Comment (st--) [reply src/LaplaceApproximationModule.jl:91]: JuliaFormatter/Reviewdog seems to have a hiccup

File: src/LaplaceApproximationModule.jl [hunk 4]
```diff
@@ -206,7 +220,7 @@ end
 `dist_y_given_f` must be a scalar function from a Real to a Distribution object.
 `ys` and `f` are vectors of observations and latent function values, respectively.
 """
-function loglik_and_derivs(dist_y_given_f, ys::AbstractVector, f::AbstractVector{<:Real})
+function loglik_and_derivs(dist_y_given_f, ys, f)
```
Comment (st--) [inline src/LaplaceApproximationModule.jl:223]:
```suggestion
function loglik_and_derivs(dist_y_given_f, ys::AbstractVector, f::AbstractVector{<:Real})
```
let's try putting this back

**Pull request merged**